src/data-access/repositories/mongo/mongo-device-repository.ts

```diff
@@ -35,15 +35,24 @@ export class MongoDeviceRepository extends BaseMongoRepository implements Device
 
   public async save(device: Device): Promise<void> {
     this.assertDatabaseConnection(MongoDeviceRepository.prototype.save.name)
+    let pureDevice: Device = {} as Device
     if (!device.id) {
-      device.id = this.uuidGenerator.generateUuid() // TODO: Remove side effect
+      pureDevice = {
+        ...device,
+        id: this.uuidGenerator.generateUuid(),
+      }
     }
```

Comment:
From the method type definition, the given device does have an `id`. If the method should support getting a device without id, the method definition should model it such that the method type becomes (e.g.) `save(device: Device | Omit<Device, 'id'>): Promise<void>`.

By doing this we know what variations of the data the method should work with. I would suggest that instead of using the check to make a imperative update, to compose the object directly. This also avoids the `{} as Device`.
```typescript
class MongoDeviceRepository {
  public save(device: Device | Omit<Device, 'id'>): Promise<void> {
    const deviceWithId: Device = {
      ...device,
      id: device.id ?? this.uuidGenerator.generateUuid(),
    }
    await this.getCollection().updateOne({ id: deviceWithId.id }, { $set: deviceWithId }, { upsert: true })
  }
}
```

Reply:
I might have overlooked something, but to me it looks like if the device has an id, then it is not saved to the collection?

Reply:
I did not notice this comment and am sorry about that. It seems like a sound best practice and I will include it.